doc/config.js.md
>**[Return to the Masterportal documentation](doc.md)**

[TOC]

# config.js

The `config.js` contains Masterportal configuration not directly related to UI or layers. For example, paths to other configuration files belong here. This file is usually placed next to the `index.html` and `config.json` files.

In the following, all configuration options are described. For all configuration options of type `object`, further nested options are linked and described in detail after the main table. You may also refer to **[this config.js example file](https://bitbucket.org/geowerkstatt-hamburg/masterportal/src/dev/portal/basic/config.js)**.

|Name|Required|Type|Default|Description|Example|
|----|--------|----|-------|-----------|-------|
|layerConf|yes|String||Path to the **[services.json](services.json.md)** file containing all available WMS layers and WFS feature types. The path is relative to *js/main.js*.|`https://geodienste.hamburg.de/lgv-config/services-internet.json"`|
|namedProjections|yes|String[]||Definition of the usable coordinate systems. See **[syntax definition](http://proj4js.org/#named-projections)** for details..|`[["EPSG:25832", "+title=ETRS89/UTM 32N +proj=utm +zone=32 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs"]]`|
|restConf|yes|String||Path to the **[rest-services.json](rest-services.json.md)** file describing further services, e.g. print service, WPS, CSW. The path is relative to *js/main.js*.|`https://geodienste.hamburg.de/lgv-config/rest-services-internet.json"`|
|styleConf|yes|String||Path to the **[style.json](style.json.md)** file describing vector layer (WFS) styles. The path is relative to *js/main.js*.|`https://geodienste.hamburg.de/lgv-config/style.json"`|
|addons|no|String[]|`[]`|List of names for custom modules. The modules are to be placed in the folder `/addons/`, with their entry points being defined in the `addonsConf.json`.|`["myAddon1", "myAddon2"]`|
|alerting|no|**[alerting](#markdown-header-alerting)**|`{"category": "alert-info", "isDismissable": true, "isConfirmable": false, "position": "top-center", "fadeOut": null}`|Overrides the alert module's default values.|{fadeOut: 6000}|
|cameraParameter|no|**[cameraParameter](#markdown-header-cameraparameter)**||Initial camera parameter||
|cesiumParameter|no|**[cesiumParameter](#markdown-header-cesiumparameter)**||Cesium flags||
|cswId|no|String|`"3"`|Reference to a CSW interface used to retrieve layer information. The ID will be resolved to a service defined in the **[rest-services.json](rest-services.json.md)** file.|`"my CSW-ID"`|
|defaultToolId|no|String|`"gfi"`|The tool with the given ID will be active when no other tool is active.|"filter"|
|layerSelector|no|**[layerSelector](#markdown-header-layerselector)**||Module to configure interactions with the layertree and the map, executed on a defined event.||
|featureViaURL|no|**[featureViaURL](#markdown-header-featureviaurl)**||Optional configuration for the URL parameter `featureViaURL`. See **[urlParameter](urlParameter.md)** for details. Implemented for treeTypes *light* and *custom*.||
|footer|no|**[footer](#markdown-header-footer)**||If set, a footer is shown and configured with this object.||
|gfiWindow|no|String|`"detached"`|_Deprecated in the next major release. Please use the attribute "Portalconfig.menu.tool.gfi.desktopType" of the **[config.json](#config.json.md)** instead._ Display type and attribute information for all layer types. **attached**: the attribute information window is opened at click position **detached**: the attribute information window is opened at the top right of the map; a marker is set to the click position.|`"attached"`|
|ignoredKeys|no|String[]|`["BOUNDEDBY", "SHAPE", "SHAPE_LENGTH", "SHAPE_AREA", "OBJECTID", "GLOBALID", "GEOMETRY", "SHP", "SHP_AREA", "SHP_LENGTH","GEOM"]`|List of attribute names to be ignored for attribute information lists of all layer types. Only used with "gfiAttributes": "showAll".|`["BOUNDEDBY", "SHAPE", "SHAPE_LENGTH", "SHAPE_AREA", "OBJECTID", "GLOBALID", "GEOMETRY", "SHP", "SHP_AREA", "SHP_LENGTH","GEOM"]`|
|infoJson|no|String|`"info.json"`|Path to the `info.json` file containing additional information on snippets. The path is relative to the *index.html*.|`"info.json"`|
|inputMap|no|Object|`{}`|If this object is set, and its field `setMarker` is set to true, the Masterportal is configured as input element for data. In that case, each click sets a Map Marker and communicates the coordinates via **[remoteInterface](remoteInterface.md)** in the chosen coordinate reference system.|`{setMarker: true, targetProjection: "EPSG:4326", setCenter: false}`|
|inputMap.setCenter|no|Boolean|`false`|Center on a marker after producing it?|`setCenter: true`|
|inputMap.setMarker|no|Boolean|`false`|Flag to activate the 'setMarker' functionality.|`setMarker: true`|
|inputMap.targetProjection|no|String|`"EPSG:25832"`|The target coordinate reference system. Coordinates will be translated to it before being communicated via **[remoteInterface](remoteInterface.md)**.|`targetprojection: "EPSG:4326"`|
|login|no|Object|`{}`|Activates the Login module. This displays a login button in the toolbar that allows login and logout with an IDM server, e.g. keycloak, as configured.||
|mapInteractions|no|**[mapInteractions](#markdown-header-mapInteractions)**||Overrides the ol map interactions. Provides further configuration possibilities for control behaviour and keyboardEventTarget.||
|mapMarker|no|**[mapMarker](#markdown-header-mapmarker)**||Overrides the map marker module's default values. Useful for 3D markers since OpenLayers's overlays can not be displayed in 3D mode. For this, the map marker has to be defined as vector layer.||
|metaDataCatalogueId|no|String|`"2"`|URL to the metadata catalog linked to in the layer information window. The ID is resolved to a service of the **[rest-services.json](rest-services.json.md)** file. Note: This attribute is only necessary, when no "show_doc_url" is configured in the metadata dataset in the **[services.json](services.json.md)**. The url can either be set globally (**[config.js](config.js.md)**) or layer-specific(**[services.json](services.json.md)**).|`"MetaDataCatalogueUrl"`|
|metadata|no|**[metadata](#markdown-header-metadata)**||Allows configuration of which metadata URLs are to be resolved via proxy.||
|mouseHover|no|**[mouseHover](#markdown-header-mousehover)**||Activates the MouseHover feature for vector layers, both WFS and GeoJSON. For per-layer configuration, see the **[config.json](config.json.md)**'s section *Themenconfig.Fachdaten.Layer*.|`true`|
|obliqueMap|no|Boolean|`false`|If set to `true`, an oblique map layer is created. An additional oblique layer must be defined.||
|portalConf|no|String|`"config.json"`|Path to the portal's `config.json` file. You may also enter a node; in that case the taken path is controlled by the urlParameter `config`.|Direct path: "../masterTree/config.json"; Node: "../../portal/master/". In the node scenario, a query parameter like `config=config.json` must exist in the URL.|
|postMessageUrl|no|String|`"http://localhost:8080"`|URL the portal is supposed to post messages to and receive messages from with the `postMessage` feature.|"http://localhost:8080"|
|proxyHost|no|String||Host name of a remote proxy with CORS configured to support the portal's domain, among others.|`"https://proxy.example.com"`|
|quickHelp|no|Object|`{}`|Activates the QuickHelp module. This displays a window containing help text for supported functions of the modules. Available for the layer tree (CustomTree), the search bar (Searchbar) and the routing tool (RoutingTool).||
|remoteInterface|no|**[remoteInterface](#markdown-header-remoteinterface)**||Optional remote interface configuration.||
|scaleLine|no|Boolean|`false`|Controls whether a scale line is displayed at the bottom right of the map. To activate scale selection (as a **ScaleSwitcher**) use the configuration parameter `isDisplayInFooter` in **[config.json](config.json)**.|`true`|
|simpleMap|no|Boolean|`false`|_Deprecated in the next major release. Please use the parameter `simpleMap` as part of the configuration of the `saveSelection` tool in the **[config.json](config.json.md)**._ Adds a SimpleMap URL to the `Save selection` dialogue. When calling this URL, the menu bar, layer tree, and map controls are deactivated. Not implemented for tree type *„light“*.|`false`|
|startingMap3D|bi|Boolean|`false`|Controls whether the map should start in 3D mode.||
|tree|no|**[tree](#tree)**||||
|uiStyle|no|String|`"default"`|Sets the control element layout. |`table`|
|wfsImgPath|no|String||Path to the folder holding images for the WFS styles. The path is relative to *js/main.js*.|`https://geodienste.hamburg.de/lgv-config/img/"`|
|wpsID|no|String|`""`|Reference to a WPS interface used in various modules. The ID is resolved to a service defined in the **[rest-services.json](rest-services.json.md)** file.|`""`|
|zoomToFeature|no|**[zoomToFeature](#markdown-header-zoomtofeature)**||_Deprecated in the next major release. Please use **[zoomTo](#markdown-header-zoomto)** instead._ Optional configuration of the URL query parameter `featureid`. For details, see **[urlParameter](urlParameter.md)**. ||
|zoomTo|no|**[zoomTo](#markdown-header-zoomto)**[]|Configuration for the URL query parameters `zoomToFeatureId` and `zoomToGeometry`.||
|layerInformation|no|**[layerInformation](#markdown-header-layerinformation)**||Configuration for the layerInformation window.||
|vuetify|no|String|undefined|Path to the optional instance of the vuetify UI library. e.g. portal or addon specific.|`addons/cosi/vuetify/index.js`|
|layerSequence|no|**[layerSequence](#markdown-header-layersequence)**||Configuration for layerSequence.||

***

## alerting

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|fetchBroadcastUrl|no|String|`false`|The alerting module will initially use a linked configuration file from this URL, if set.|
|localStorageDisplayedAlertsKey|no|String|`"displayedAlerts"`|Arbitrary key used to store information regarding the alerting module in the browser's local storage.|

***

## cameraParameter

Cesium Scene camera settings in 3D mode.
_Deprecated in the next major release. Please use **[cesiumParameter](#markdown-header-cesiumParameter)** instead._

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|heading|no|Number||Camera's initial heading in radians|
|tilt|no|Number||Camera's initial tile in radians|
|altitude|no|Number||Camera's initial height in meters|

***

## cesiumParameter

Cesium Scene settings in 3D mode.
For more attributes see **[Scene](https://cesium.com/learn/cesiumjs/ref-doc/Scene.html?classFilter=scene)**

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|camera|no|**[camera](#markdown-header-cesiumParametercamera)**||Cesium Scene camera settings in 3D mode.|
|fog|no|**[fog](#markdown-header-cesiumParameterfog)**||Cesium Scene fog settings in 3D mode.|
|fxaa|no|Boolean|`true`|activates *fast approximate anti-aliasing*|
|globe|no|**[globe](#markdown-header-cesiumParameterglobe)**||Cesium Scene globe settings in 3D mode.|
|maximumScreenSpaceError|no|Number|`2.0`|Detail level in which terrain/raster tiles are fetched. 4/3 is the highest quality level.|
|tileCacheSize|no|Number|`100`|terrain/raster tile cache size|

**Example**

```json
{
    "camera": {
        "altitude": 127,
        "heading": -1.2502079000000208,
        "tilt": 45
    },
    "fog": {
        "enabled": true
    },
    "fxaa": true,
    "globe": {
        "enableLighting": true
    },
    "maximumScreenSpaceError": 2,
    "tileCacheSize": 20,
}
```
***

### cesiumParameter.camera

Cesium Scene camera settings in 3D mode.
The camera is defined by a position, orientation, and view frustum.
For more attributes see **[Scene](https://cesium.com/learn/cesiumjs/ref-doc/Camera.html)**

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|altitude|no|Number||Camera's initial height in meters|
|heading|no|Number||Camera's initial heading in radians|
|tilt|no|Number||Camera's initial tile in radians|

**Example**

```json
{
    "camera": {
        "altitude": 127,
        "heading": -1.2502079000000208,
        "tilt": 45
    }
}
```
***

### cesiumParameter.fog

Cesium Scene fog settings in 3D mode.
Blends the atmosphere to geometry far from the camera for horizon views.
For more attributes see **[Scene](https://cesium.com/learn/cesiumjs/ref-doc/Fog.html)**

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|enabled|no|Boolean|`false`|True if fog is enabled.|

**Example**

```json
{
    "fog": {
        "enabled": true
    }
}
```
***

### cesiumParameter.globe

Cesium Scene globe settings in 3D mode.
The globe rendered in the scene, including its terrain and imagery layers.
For more attributes see **[Scene](https://cesium.com/learn/cesiumjs/ref-doc/Globe.html)**

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|enableLighting|no|Boolean|`false`|Activates light effects on the map based on the sun's position.|

**Example**

```json
{
    "globe": {
        "enableLighting": true
    }
}
```
***

## footer

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|urls|no|**[urls](#markdown-header-footerurls)**[]||Array of URL configuration objects.|
|showVersion|no|Boolean|`false`|If `true`, the Masterportal version number is included in the footer.|
|footerInfo|no|**[footerInfo](#markdown-header-footerfooterInfo)**[]||Array of information configuration objects.|
|mobileFooterInfoToggler|no|Boolean|`false`|If `true`, it is possible to toggle the footer info in mobile version.|

***

### footer.urls
|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|alias|no|String|`"Landesbetrieb Geoinformation und Vermessung"`|Link text for desktop applications|
|alias_mobil|no|String|`"LGV"`|Link text for mobile applications|
|bezeichnung|no|String|`"Kartographie und Gestaltung: "`|Link prefix|
|url|no|String||the URL the `alias` (or `alias_mobile`) text links to, e.g. `"https://example.com/"`. You may also link to mail creation with e.g. `"mailto:my_mail@example.com"`|
|toolModelId|no|String|`"SdpDownload"`|The id of a module to be opened on clicking the link. Do not include a URL for such cases.|

**Example:**

```json
{
    "footer": {
        "urls": [
            {
                "bezeichnung": "Cartography and design: ",
                "url": "https://geoinfo.hamburg.de/",
                "alias": "Landesbetrieb Geoniformation und Vermessung",
                "alias_mobil": "LGV"
            },
            {
                "bezeichnung": "",
                "url": "http://www.hamburg.de/bsu/timonline",
                "alias": "Map inconsistencies"
            },
            {
                "bezeichnung": "",
                "url": "",
                "alias": "SDP Download",
                "toolModelId": "sdpdownload"
            }
        ],
        "showVersion": true
    }
}
```

***

### footer.footerInfo
|Name|Required|Type|Default|Description|
|----|-------------|---|-------|------------|
|title|yes|String||Title of the information tab.|
|description|no|String||Text displayed under the title.|
|subtexts|no|**[subtexts](#markdown-header-footerfooterInfosubtexts)**[]||Array of subtext configuration objects.|

***

### footer.footerInfo.subtexts
|Name|Required|Type|Default|Description|
|----|-------------|---|-------|------------|
|subtitle|yes|String||Subtitle of the subtext.|
|text|no|String||Text displayed under the subtitle.|

**Beispiel:**

```json
{
    "footerInfo": [
        {
            "title": "Contact",
            "description": "Text under the titel",
            "subtexts": [
                {
                    "subtitle": "Postal address",
                    "text": "Max-Mustermann-Str. 1 <br> 12345 City <br> Germany"
                },
                {
                    "subtitle": "Phone and fax",
                    "text": "Tel: +49 (0) 1234 56789 <br> Fax: +49 (0) 1234 5678910"
                }
            ]
        },
        {
            "title": "Privacy",
            "subtexts": [
                {
                    "subtitle": "Subtitle",
                    "text": "Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt..."
                }
            ]
        }
    ]
}
```
***

## login

This module allows the user to login with an OIDC server. The retrieved access token is stored in cookies which can be used by the backend to deliver user-specific data (e.g. layers). Since the cookies are technically required to implement the login functionality, there is not corresponding cookie notice.

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|oidcAuthorizationEndpoint|yes|String||The oidc auth endpoint, e.g. "https://idm.domain.de/auth/realms/REALM/protocol/openid-connect/auth".|
|oidcRevocationEndpoint|yes|String||The oidc revoke endpoint, e.g. "https://idm.domain.de/auth/realms/REALM/protocol/openid-connect/revoke".|
|oidcTokenEndpoint|yes|String||The oidc token endpoint, e.g. "https://idm.domain.de/auth/realms/REALM/protocol/openid-connect/token".|
|oidcClientId|yes|String||The oidc client, e.g. "masterportal" (must be created in your IDM, e.g. keycloak).|
|oidcScope|yes|String||The scope used for oidc, defaults to "profile email openid".|
|oidcRedirectUri|yes|String||The url to redirect the oidc process to - after login.|
|interceptorUrlRegex|yes|String||An regexp pattern that allows to specify urls the oidc token will be attached to.|

Make sure in keycloak the client is configured as follows:

```
Access Type: public
Standard Flow Enabled: ON
Valid Redirect URIs: <PORTAL URL, e.g. https://localhost/*>
Web Origins: <PORTAL HOST, e.g. https://localhost>
```

Specific ports are not allowed ports here. Especially, ports on localhost, e.g. `localhost:9001` will not work with keycloak since ports are not accepted in web origins.

In the section `OpenID Connect Compatibility Modes` activate `Use Refresh Tokens`. In the section `Advanced Settings` set `Proof Key for Code Exchange Code Challenge Method` to `S256`.

**Example:**

```json
{
  "login": {
      "oidcAuthorizationEndpoint": "https://idm.DOMAIN.de/auth/realms/REALM/protocol/openid-connect/auth",
      "oidcRevocationEndpoint": "https://idm.DOMAIN.de/auth/realms/REALM/protocol/openid-connect/revoke",
      "oidcTokenEndpoint": "https://idm.DOMAIN.de/auth/realms/REALM/protocol/openid-connect/token",
      "oidcClientId": "masterportal",
      "oidcScope": "profile email openid",
      "oidcRedirectUri": "https://localhost/portal/basic/",
      "interceptorUrlRegex": "https?://localhost.*" // add authorization to all URLs that match the given regex
  }
}
```

## mapInteractions

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|interactionModes|no|Object|`{dragPan: false, altShiftDragRotate: false, pinchRotate: false}`| Interaction settings for the ol default interactions. If not set, the default setting is used.|
|keyboardEventTarget|no|Object/false|`false`|possibility to set the keyboard event target for the ol map e.g keyboardEventTarget: document|

**Example:**

```json
{
    "mapInteractions": {
        "interactionModes": {"dragPan": false, "altShiftDragRotate": true, "pinchRotate": false, "dragZoom": true},
        "keyboardEventTarget": false
    }
}
```

***
## mapMarker

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|pointStyleId|no|String|`"defaultMapMarkerPoint"`|StyleId to refer to a `style.json` point style. If not set, the `img/mapMarker.svg` is used.|
|polygonStyleId|no|String|`"defaultMapMarkerPolygon"`|StyleId to refer to a `style.json` polygon style.|
|additionalPolygonStyleId|no|String|`"defaultAdditionalMapMarkerPolygon"`| StyleId to refer to an additional `style.json` polygon style.|

**Example:**

```json
{
    "mapMarker": {
        "pointStyleId": "customMapMarkerPoint",
        "polygonStyleId": "customMapMarkerPolygon",
        "additionalPolygonStyleId": "customAdditionalMapMarkerPolygon"
    }
}
```

***

## mouseHover

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|minShift|no|Integer|`5`|Minimum mouse position movement required to render a new tooltip; in pixels.|
|numFeaturesToShow|no|Integer|`2`|Maximum amount of element information per tooltip; when exceeded, an information text informs the user of cut content.|
|infoText|no|String|`"(Further objects. Please zoom.)"`|Information text shown when `numFeaturesToShow` is exceeded.|
|highlightOnHover|no|Boolean|false|If hovered features should be highlighted|
|highlightVectorRulesPolygon|no|Object||Specify the fill color and outline color and stroke width for highlighting the polygon features as well as a zoom parameter.|
|highlightVectorRulesPointLine|no|Object||Specify outline color and stroke width for highlighting lines and fill color and scale factor for highlighting points as well as a zoom parameter.|

**Example:**

```json
{
    "mouseHover": {
        "numFeaturesToShow": 2,
        "infoText": "The Info Text",
        "highlightOnHover": true,
        "highlightVectorRulesPolygon": {
            "fill": {
                "color": [255, 255, 255, 0.5]
            },
            "stroke": {
                "width": 4,
                "color": [255, 0, 0, 0.9]
            }
        },
        "highlightVectorRulesPointLine": {
            "stroke": {
                "width": 8,
                "color": [255, 0, 255, 0.9]
            },
            "image": {
                "scale": 2
            }
        }
    },
}
```

### mouseHover.highlightVectorRulesPolygon

Specify the fill color and outline color and stroke width for highlighting the polygon features as well as a zoom level.

|Name|Required|Type|Default|Description|Expert|
|----|--------|----|-------|-----------|------|
|fill|no|**[fill](#markdown-header-mousehoverhighlightvectorrulespolygonfill)**||Possible setting: color|false|
|stroke|no|**[stroke](#markdown-header-mousehoverhighlightvectorrulespolygonstroke)**||Possible setting: width|false|

***

### mouseHover.highlightVectorRulesPolygon.fill
|Name|Required|Type|Default|Description|Expert|
|----|--------|----|-------|-----------|------|
|color|no|Float[]|[255, 255, 255, 0.5]|Possible setting: color (RGBA)|false|

```json
"fill": { "color": [215, 102, 41, 0.9] }
```

***

### mouseHover.highlightVectorRulesPolygon.stroke
|Name|Required|Type|Default|Description|Expert|
|----|--------|----|-------|-----------|------|
|width|no|Integer|1|Possible setting: width|false|
|color|no|Float[]|[255, 0, 0, 0.9]|Possible setting: color (RGBA)|false|

```json
"stroke": { "width": 4 , "color": [255, 0, 255, 0.9]}
```

***


### mouseHover.highlightVectorRulesPointLine

Specify outline color and stroke width for highlighting lines and fill color and scale factor for highlighting points. Also a zoom level.

|Name|Required|Type|Default|Description|Expert|
|----|--------|----|-------|-----------|------|
|fill|no|**[fill](#markdown-header-mousehoverhighlightvectorrulespointlinefill)**||Possible setting: color|false|
|stroke|no|**[stroke](#markdown-header-mousehoverhighlightvectorrulespointlinestroke)**||Possible setting: width|false|
|image|no|**[image](#markdown-header-mousehoverhighlightvectorrulespointlineimage)**||Possible setting: scale|false|

***
### mouseHover.highlightVectorRulesPointLine.fill
|Name|Required|Type|Default|Description|Expert|
|----|--------|----|-------|-----------|------|
|color|no|Float[]|[255, 255, 255, 0.5]|Possible setting: color (RGBA)|false|

```json
"fill": { "color": [215, 102, 41, 0.9] }
```

***

### mouseHover.highlightVectorRulesPointLine.stroke
|Name|Required|Type|Default|Description|Expert|
|----|-------------|---|-------|------------|------|
|width|no|Integer|1|Possible setting: width|false|
|color|no|Float[]|[255, 255, 255, 0.5]|Possible setting: color (RGBA)|false|

```json
"stroke": { "width": 4 , "color": [255, 0, 255, 0.9]}
```

***

### mouseHover.highlightVectorRulesPointLine.image
|Name|Required|Type|Default|Description|Expert|
|----|--------|----|-------|-----------|------|
|scale|no|Integer|1.5|Possible setting: scale|false|

```json
"image": { "scale": 2}
```

***

## portalLanguage

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|enabled|yes|Boolean|`true`|Controls whether a button to switch the portal's language is provided.|
|debug|no|Boolean|`false`|Controls whether debug information regarding translations is logged to the console.|
|languages|yes|Object|`{ de: "deutsch", en: "englisch" }`|Language abbreviations. Please mind that matching locale files must exist.|
|fallbackLanguage|no|String|`"de"`|Fallback language used if contents are not available in the currently selected language.|
|changeLanguageOnStartWhen|no|String[]|`["querystring", "localStorage", "navigator", "htmlTag"]`|Order of user language detection. See [i18next browser language detection documentation](https://github.com/i18next/i18next-browser-languageDetector) for details.|
|loadPath|no|String|`"/locales/{{lng}}/{{ns}}.json"`|Path to load language files from, or a function returning such a path: `function(lngs, namespaces) { return path; }`. `lng` and `ns` are read from the path, if given, as if from a static path. You may also provide a URL like `"https://localhost:9001/locales/{{lng}}/{{ns}}.json"`. See [i18next http backend documentation](https://github.com/i18next/i18next-http-backend) for details.|

**Example:**

```json
{
"portalLanguage": {
        "enabled": true,
        "debug": false,

CHANGELOG.md
# Changelog Masterportal
 All important changes in this project are stored in this file.

[Semantic versioning](https://semver.org/spec/v2.0.0.html) is used.

## Unreleased - in development
### __Breaking Changes__
- Searchbar config.json configuration: If a search interface is used in topic tree searchbar, new parameter 'searchInterfaceId' must be filled and used in topic tree searchbar configuration.
- Searchbar in topic tree config.json configuration: 'searchInterfaceInstances' must contain the search interfaces to use. 'searchInterfaceInstanceIds' and 'searchCategory' have been removed.

### Added
- AddWMS: Added Information about CORS header to doc and changed error message.
- ShareView: Error handling when sharing via QR code has been improved.
- Maps: add filter to getFeature Request in zoomToGetAndFilterFeatures function.
- Routing: Added checkbox to determine if routes stay visible after closing the routing tool
- Login: Documentation was added to [config.js](https://bitbucket.org/geowerkstatt-hamburg/masterportal/src/dev_vue/doc/config.js.md). 

### Changed

### Deprecated

### Removed
- Deleted the `namespaces` option, as it is not a valid configuration for `createStore`, and also removed it from the `docs/Dev/unitTestVue.md` tutorial.

### Fixed
- Fixed the issue that the legend information was empty when an empty legend-object was passed to the legend component.
- Corrected link for geodata infrastructure information pdf by GDI-DE. 
- Issue #1276: Duplicate icons StartModule in expandable ControlBar.
- Print: Fixed opacity for wfs group layers
- `tree.type: auto`: layers configured in config.json do overwrite configuration in services.json.

---

## 2024-10-09 v3.3.1 (LTS)

### Fixed
- Reduced loading requests for layers defined as array of Ids (not of typ GROUP).
- Print: Fix printing bug considering minScale and maxScale with layers defined as array of ids.
- LayerInformation: Resolved an issue where the tab appeared due to layerInfo.url being undefined instead of null in the layerInformation component.

---

## 2024-10-02 v3.3.0 (LTS)

### Added
- Release Schedule and release information to readme.md.
- FileImport: Added the possibility to configure the confirmation window to show or not.
- Searchbar: mapMarker or polygonMarker stays visible on close of search results.

### Fixed
- AlertingItem: Fixed the position of closing button.
- Issue #1166: function isHTML now recognizes more HTML tags in one String as valid.
- Issue #1179: Add Parameter "WithCredentials" to highlightFeaturesByAttribute axios request, so that it works with secured layers.
- LayerInfoContactButton: no warning in console, if at layer `md_name` ist not available in `datasets`.
- LayerSelection: `tree.type: auto`: change of category creates correct folders.
- Draw_old: import KML file works correctly.
- GroupLayer: Layer of the type group consider now min and max scale of the single layer in 2D.

---

## 2024-09-19 v3.2.2
### Fixed
- 3D: `tree.type: auto` 3D layers can also be selected in the subject tree if additional layers are configured in config.json under `subjectlayer`.

---

## 2024-09-12 v3.2.1
### Fixed
- UrlParams:
    - mapMarker: Fixed map marker set by url parameter.
    - layerIds: if a layerId in params is part of a baselayer group, the group is loaded.

---

## 2024-09-04 v3.2.0
### Added
- WFST: Added buttons names for confirm and stop.
- Alerting: Examples for the configuration of messages have been added to the alerting documentation.
- Modeler3D: Modeler3D was added
- LayerTree:
    - In SubMenu added button to open contact form with layer specific parameters to be used in email.
    - Adds parameter buttonTitle in addLayerButton to enable to configure the button title.

### Changed
- LayerSelection
    - Switches layer categories selection from <select.. to <radio.. using a collapsed list open/close within a IconButton.
    - Adds handling for parameters in `config.json`/`portalConfig.tree`:
        - `hideBackgroundsHeader` to hide the headline for backgrounds
        - `backgroundsHeaderText` custom headline for the backgrounds (to overwrite the default translation)
        - `hideDatalayerHeader` to hide the headline for datalayers
        - `datalayerHeaderText` custom headline for the datalayers (to overwrite the default translation)
- Login:
    - Access Tokens are now revoked after logout.
    - The WMS-Time layer are now available after you have logged in with the login module.
- The following packages have been updated:
    - dependencies:
        - @masterportal/masterportalapi: 2.39.0 to 2.40.0
        - axios: 1.7.1 to 1.7.7
        - chart.js: 4.4.3 to 4.4.4
        - i18next: 23.11.5 to 23.15.1
        - i18next-http-backend: 2.5.2 to 2.6.1
        - qrcode: 1.5.3 to 1.5.4
        - three: 0.165.0 to 0.168.0
    - devDependencies:
        - @eslint/js: 9.3.3 to 9.10.0
        - @geoblocks/print: 0.7.8 to 0.7.9
        - @sinonjs/fake-timers: 11.2.2 to 13.0.1
        - @stylistic/eslint-plugin-js: 2.2.1 to 2.8.0
        - @vue/compiler-sfc: 3.3.4 to 3.5.4
        - @vue/devtools-api: 6.5.0 to 6.6.4
        - @vue/test-utils: 2.4.1 to 2.4.6
        - chai: 4.4.1 to 5.1.1
        - eslint-plugin-jsdoc: 48.2.7 to 50.2.2
        - eslint-plugin-mocha: 10.4.3 to 10.5.0
        - eslint-plugin-n: 17.9.0 to 17.10.2
        - eslint-plugin-vue: 9.25.0 to 9.28.0
        - eslint-plugin-vuejs-accessibility: 2.3.0 to 2.4.1
        - globals: 15.4.0 to 15.9.0
        - mocha: 10.4.0 to 10.7.3
        - sinon-chai: 3.7.0 to 4.0.0
        - vue-loader: 17.3.0 to 17.4.2
        - zip-a-folder: 3.1.6 to 3.1.7
- Import GeoJson
    - GFI attributes are added in the import of a standard GeoJson
    - Supports a new structure of the draw style properties which are set in the export of the draw tool as well as the old ones
    - Adds a custom styling option for GeoJson/ Json import

portal/master/config.js
const Config = {
    addons: ["bildungsatlas", "dataTable", "dipas", "exampleSearch", "populationRequest", "schulinfo", "solaratlas", "sturmflut", "trafficCount", "trinkwasser", "verkehrsstaerken"],
    ignoredKeys: ["BOUNDEDBY", "SHAPE", "SHAPE_LENGTH", "SHAPE_AREA", "OBJECTID", "GLOBALID", "GEOMETRY", "SHP", "SHP_AREA", "SHP_LENGTH", "GEOM"],
    namedProjections: [
        // ETRS89 UTM
        ["EPSG:4647", "+title=ETRS89_UTM32, EPSG 4647 (zE-N) +proj=tmerc +lat_0=0 +lon_0=9 +k=0.9996 +x_0=32500000 +y_0=0 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs +type=crs"],
        // EPSG: 25832
        ["EPSG:25832", "+title=EPSG 25832 +proj=utm +zone=32 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs"],
        // ETRS89_3GK3
        ["EPSG:8395", "+title=ETRS89_3GK3 +proj=tmerc +lat_0=0 +lon_0=9 +k=1 +x_0=3500000 +y_0=0 +ellps=GRS80 +datum=GRS80 +units=m +no_defs"],
        // EPSG: 8395
        ["EPSG:8395", "+title=EPSG 8395 +proj=tmerc +lat_0=0 +lon_0=9 +k=1 +x_0=500000 +y_0=0 +ellps=GRS80 +datum=GRS80 +units=m +no_defs"],
        // GK DHDN
        ["EPSG:31467", "+title=DE_DHDN_3GK3, EPSG 31467 +proj=tmerc +lat_0=0 +lon_0=9 +k=1 +x_0=3500000 +y_0=0 +ellps=bessel +datum=potsdam +units=m +no_defs"],
        // WGS84
        ["EPSG:4326", "+title=WGS84_Lat-Lon (Grad, Minuten, Sekunden), EPSG 4326 +proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs"]
    ],
    layerConf: "https://geodienste.hamburg.de/services-internet.json",
    restConf: "https://geodienste.hamburg.de/lgv-config/rest-services-internet.json",
    styleConf: "https://geodienste.hamburg.de/lgv-config/style_v3.json",
    wfsImgPath: "https://geodienste.hamburg.de/lgv-config/img/",
    alerting: {
        fetchBroadcastUrl: "./resources/newsFeedPortalAlerts.json",
        initialAlerts: {
            "qs-release": {
                category: "Portal zur Abnahme!",
                content: "Dieses Geoportal dient der Qualitätskontrolle durch den Kunden.<br>Es ist aufgrund von möglichen Fehlern <b>nicht</b> zur Nutzung für alltägliche oder berufliche Aufgaben geeignet!<br><br>",
                creationDate: "01/09/22",
                mustBeConfirmed: true,
                once: true
            }
        }
    },
    cesiumLibrary: "https://geoportal-hamburg.de/mastercode/cesium/latest/Cesium.js",
    portalLanguage: {
        enabled: true,
        debug: false,
        languages: {
            de: "Deutsch",
            en: "English",
            es: "Español",
            it: "Italiano",
            platt: "Platt",
            pt: "Português",
            ru: "Русский",
            tr: "Türkçe",
            ua: "Українська"
        },
        fallbackLanguage: "de",
        changeLanguageOnStartWhen: ["querystring", "localStorage", "htmlTag"]
    }
};

if (typeof module !== "undefined") {
    module.exports = Config;
}
